treeDetectLmax.py

```diff
@@ -0,0 +1,276 @@
+        geometry.Destroy()
+        feature.Destroy()
+    # Cleanup
+    shapeData.Destroy()
+    # Return
+    return shapePath
+
+
+def kmlsave(kmlPath, trees):
+    ''' Save data to a kml file
+        @param  kmlPath       Input path to the kml which will be created (Str)
+        @param  trees         Input tree data (dictionnary)
+    '''
+    geoLocations = np.column_stack((trees['xpos'],trees['ypos']))
+    height = trees['height']
+    
+    driver = ogr.GetDriverByName('KML')
+
+    # Create shapeData
+    kmlPath = validateKmlPath(kmlPath)
+    if os.path.exists(kmlPath):
+        os.remove(kmlPath)
+        
+    kmlData = driver.CreateDataSource(kmlPath)
+    # Create spatialReference
+    outSpatialRef = osr.SpatialReference()
+    outSpatialRef.ImportFromEPSG(4326) # output SpatialReference
+    
+    spatialReference = getSpatialReferenceFromWkt(trees['prj_wkt'])
+    transform = osr.CoordinateTransformation(spatialReference, outSpatialRef)
+
+    # Create layer
+    layerName = os.path.splitext(os.path.split(kmlPath)[1])[0]
+    layer = kmlData.CreateLayer(layerName, outSpatialRef, ogr.wkbPoint)
+    
+    field_height = ogr.FieldDefn("H", ogr.OFTReal)
+    field_height.SetWidth(4)
+    field_height.SetPrecision(2)
+    
+    layer.CreateField(field_height)
+    
+    for pointIndex, geoLocation in enumerate(geoLocations):
+        # Create point
+        geometry = ogr.Geometry(ogr.wkbPoint)
+        geometry.SetPoint(0, geoLocation[0], geoLocation[1])
+        geometry.Transform(transform)
+        
+        # Create feature
+        feature = ogr.Feature(layer.GetLayerDefn())
+        feature.SetGeometry(geometry)
+        feature.SetFID(pointIndex)
+        feature.SetField("H",float(height[pointIndex]))
+        
+        # Save feature
+        layer.CreateFeature(feature)
+        # Cleanup
+        geometry.Destroy()
+        feature.Destroy()
+    # Cleanup
+    kmlData.Destroy()
+    # Return
+    return kmlPath
+    
+
+def getSpatialReferenceFromWkt(Wkt):
+    '''Return GDAL spatial reference object from proj4 string'''
+    spatialReference = osr.SpatialReference()
+    spatialReference.ImportFromWkt(Wkt)
+    return spatialReference
+    
+def validateShapePath(shapePath):
+    '''Validate shapefile extension'''
+    return os.path.splitext(str(shapePath))[0] + '.shp'
+    
+def validateKmlPath(kmlPath):
+    '''Validate kml extension'''
+    return os.path.splitext(str(kmlPath))[0] + '.kml'    
+    
+# convert pixel coordinates to geospatial coordinates
+def ApplyGeoTransform(inx,iny,gt):
+    ''' Apply a geotransform
+        @param  inx       Input x coordinate (double)
+        @param  iny       Input y coordinate (double)
+        @param  gt        Input geotransform (six doubles)
+        @return outx,outy Output coordinates (two doubles)
+    '''
+    outx = gt[0] + inx*gt[1] + iny*gt[2]
+    outy = gt[3] + inx*gt[4] + iny*gt[5]
+    return (outx,outy)    
+
+
+# if __name__ == "__main__":
+    
```

Comment:
My question was more about:

Si on rajoute :

if __name__ == "__main__":
    main()

dans son script pour appeler la fonction « main » qu’on a définie plus haut,

1.	J’ai lu que c’était bien car ça évitait l’exécution de la fonction lors d’un import (vrai ?)

2.	Que cette partie ne s’activait que lors d’un appel par une ligne de commande (vrai ?)

2.bisEst-ce qu’on peut activer cette partie à partir d’un autre script ? (Dans un autre fichier)

But I answered it with http://enacit1.epfl.ch/introduction-python/#écriture-de-modules

@monodo about classes, I think that we can clarify the code structure by using a class for the geospatial methods like applygeotransform, ... which will be used multiple times and I agree that methods (which are processing steps like treeDetectLMax, crownDetection, ... will be children of the forLim class)